zcash_primitives/src/transaction/builder.rs

```diff
@@ -323,32 +496,16 @@ impl<'a, P: consensus::Parameters, R: RngCore> Builder<'a, P, R> {
         // Consistency checks
         //
 
-        // Valid change
-        let change = (self.value_balance()? - self.fee).ok_or(Error::InvalidAmount)?;
-
-        if change.is_negative() {
-            return Err(Error::ChangeIsNegative(change));
+        // No additional change must be required in order to balance.
+        if balances.change_required() > Amount::zero() {
+            return Err(Error::ChangeRequired(balances.change_required()));
         }
 
-        //
-        // Change output
-        //
-
-        if change.is_positive() {
-            // Send change to the specified change address. If no change address
-            // was set, send change to the first Sapling address given as input.
-            match self.change_address.take() {
-                Some(ChangeAddress::SaplingChangeAddress(ovk, addr)) => {
-                    self.add_sapling_output(Some(ovk), addr, change, MemoBytes::empty())?;
-                }
-                None => {
-                    let (ovk, addr) = self
-                        .sapling_builder
-                        .get_candidate_change_address()
-                        .ok_or(Error::NoChangeAddress)?;
-                    self.add_sapling_output(Some(ovk), addr, change, MemoBytes::empty())?;
-                }
-            }
+        // After fees are accounted for, the value balance of the transaction must be zero.
```

Comment:
Add a comment about the apparent redundancy between this check and the one for change above.